cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/AnomalyDetectorState.java

```diff
@@ -62,6 +64,19 @@ protected boolean removeEldestEntry(Map.Entry<String, AnomalyState> eldest) {
       });
     }
     _selfHealingEnabled = selfHealingEnabled;
+    _selfHealingEnabledRatio = null;
+  }
+
+  void setSelfHealingEnabledRatio(Map<AnomalyType, Float> selfHealingEnabledRatio) {
+    if (selfHealingEnabledRatio == null) {
+      throw new IllegalArgumentException("Attempt to set selfHealingEnabledRatio with null.");
+    }
+    _selfHealingEnabledRatio = new HashMap<>(selfHealingEnabledRatio.size());
+    selfHealingEnabledRatio.forEach((key, value) -> _selfHealingEnabledRatio.put(key.name(), value));
+  }
+
+  private Map<String, Float> selfHealingEnabledRatio() {
+    return _selfHealingEnabledRatio == null ? Collections.emptyMap() : _selfHealingEnabledRatio;
```

Comment:
looks like `_selfHealingEnabledRatio == null` will always be `false` since AnomalyDetector [always set ratio before turn AnomalyDetectorState](https://github.com/linkedin/cruise-control/blob/6d8d98f90dab4ec473bc6b8f868b4d1b8bceb340/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/AnomalyDetector.java#L187)?

Reply:
I think we should be careful to distinguish (1) the existing implementation (how the functions are currently accessed) and (2) supported use-cases (how the functions can be accessed) -- i.e. please see [defensive programming](https://en.wikipedia.org/wiki/Defensive_programming).

This check provides protection against an `NPE` when/if there is an attempt to access the public interfaces (1) `getJsonStructure()` or (2) `toString()` without setting self healing enabled ratio. A potential improvement that I see is to make `setSelfHealingEnabledRatio` public rather than package private to provide it with the same access constraints as  `getJsonStructure()` and `toString()`.